service/controller/v15/resource/drainer/create.go

```diff
@@ -0,0 +1,183 @@
+package drainer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	corev1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/core/v1alpha1"
+	providerv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/microerror"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/aws-operator/service/controller/v15/controllercontext"
+	"github.com/giantswarm/aws-operator/service/controller/v15/key"
+)
+
+// EnsureCreated tries to drain guest cluster nodes when necessary. Once it
+// detects a guest cluster node being in terminating/wait state in EC2 a
+// DrainerConfig is created to instruct the node-operator to drain the specific
+// node. The node-operator updates the DrainerConfig state as soon as it has
+// drained the node and the aws-operator here completes the lifecycle hook of
+// the drained node and deletes the DrainerConfig.
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	controllerCtx, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	workerASGName := controllerCtx.Status.Drainer.WorkerASGName
+	if workerASGName == "" {
+		// TODO add log fmt entry about controllercontext.Context.Status values retrival.
+		r.logger.LogCtx(ctx, "level", "debug", "message", "worker ASG name is not available yet")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+		return nil
+	}
+
+	var instances []*autoscaling.Instance
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding out the guest cluster nodes being in state '%s'", autoscaling.LifecycleStateTerminatingWait))
+
+		i := &autoscaling.DescribeAutoScalingGroupsInput{
+			AutoScalingGroupNames: []*string{
+				aws.String(workerASGName),
+			},
+		}
+
+		o, err := controllerCtx.AWSClient.AutoScaling.DescribeAutoScalingGroups(i)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		for _, g := range o.AutoScalingGroups {
+			for _, i := range g.Instances {
+				if *i.LifecycleState == autoscaling.LifecycleStateTerminatingWait {
+					instances = append(instances, i)
+				}
+			}
+		}
+
+		if len(instances) == 0 {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("did not find the guest cluster nodes being in state '%s'", autoscaling.LifecycleStateTerminatingWait))
+			r.logger.LogCtx(ctx, "level", "debug", "message", "nothing to do")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			return nil
```

Comment:
Again when resources are split and have single responsibility we can cancel early. `nothing to do` is just a reason log for those who are not familiar with the codebase, and for those who forget quickly (like myself).